core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -84,102 +77,70 @@ class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
         RequestAndCompletionHandler(brokerRequestQueue.node, new WriteTxnMarkersRequest.Builder(markersToSend), requestCompletionHandler)
       }
     }
   }
 
-
   def addOrUpdateBroker(broker: Node) {
     brokerStateMap.putIfAbsent(broker.id(), new BrokerRequestQueue(broker)) match {
       case Some(brokerQueue) => brokerQueue.maybeUpdateNode(broker)
       case None => // nothing to do
     }
   }
 
-  private[transaction] def addRequestForBroker(brokerId: Int, metadataPartition: Int, txnMarkerEntry: TxnMarkerEntry) {
-    val brokerQueue = brokerStateMap(brokerId)
-    brokerQueue.addRequests(metadataPartition, txnMarkerEntry)
-    trace(s"Added marker $txnMarkerEntry for broker $brokerId")
-  }
+  def addTxnMarkersToSend(transactionalId: String, pid: Long, epoch: Short, result: TransactionResult, coordinatorEpoch: Int, topicPartitions: immutable.Set[TopicPartition]): Unit = {
+    val txnTopicPartition = txnStateManager.partitionFor(transactionalId)
+    val partitionsByDestination: immutable.Map[Node, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
+      var brokerNode: Option[Node] = None
 
-  def addRequestToSend(metadataPartition: Int, pid: Long, epoch: Short, result: TransactionResult, coordinatorEpoch: Int, topicPartitions: immutable.Set[TopicPartition]): Unit = {
-    val partitionsByDestination: immutable.Map[Int, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
-      val currentBrokers = mutable.Set.empty[Int]
-      var brokerId:Option[Int] = None
-
-      while(brokerId.isEmpty) {
-        val leaderForPartition = metadataCache.getPartitionInfo(topicPartition.topic, topicPartition.partition)
-        leaderForPartition match {
-          case Some(partitionInfo) =>
-            val leaderId = partitionInfo.leaderIsrAndControllerEpoch.leaderAndIsr.leader
-            if (currentBrokers.add(leaderId)) {
-              try {
-                metadataCache.getAliveEndpoint(leaderId, interBrokerListenerName) match {
-                  case Some(broker) =>
-                    addOrUpdateBroker(broker)
-                    brokerId = Some(leaderId)
-                  case None =>
-                    currentBrokers.remove(leaderId)
-                    trace(s"alive endpoint for broker with id: $leaderId not available. retrying")
-
-                }
-              } catch {
-                case _:BrokerEndPointNotAvailableException =>
-                  currentBrokers.remove(leaderId)
-                  trace(s"alive endpoint for broker with id: $leaderId not available. retrying")
-              }
-            }
-          case None =>
-            trace(s"couldn't find leader for partition: $topicPartition")
-        }
-        if (brokerId.isEmpty)
+      while (brokerNode.isEmpty) {
```

Comment:
yeah I have thought about that too.. The approach is that if we cannot locate the broker to offer to its queue then in order for this thread to return we likely need to put it in an "unknown broker" queue first, and periodically we can check the queue and migrate to the found brokers. Does that make sense?